### DISCUSSION TITLE
The server does not receive the message when it has successfully sent the message and closes the application immediately

### DISCUSSION THREAD

**Discussion by binbon2d** [jchristn/WatsonTcp#243] [Q&A]
I have a problem in implementing a Client Console to send messages to the Server
As soon as the Console is connected to the Server I send a message and close the Console and the bad thing happened, the Server couldn't receive my message.
To solve this problem I had to let the console sleep for 500ms before closing it.
I want to ask the cause and how to fix it effectively, I can't keep my Console open all the time.

**Comment by jchristn** [comment]
Hi @binbon2d could you share a small code sample that highlights this problem?  An alternative would be to use a ```SendAndWait``` method.

**Reply by binbon2d** [reply]
Very simple, 
1. Open Console
2. Connect Server
3. Send Message
4. Close Consoel

static void Main(string[] args)
        {
            try
            {
                
                for (int i = 0; i < args.Length; i++)
                {
                    NLogLogger.Info(new string[] { "TCPCommand", "Main", "Receive", $"Arg[{i}] = [{args[i]}]" });
                    if (i == args.Length - 1)
                    {
                        NLogLogger.Info(new string[] { "TCPCommand", "Main", "END Receive ------------------------------------" });
                    }
                }
                client = new WatsonTcpClient("192.168.1.11", 9000);
                client.Events.ServerConnected += ServerConnected;
                client.Events.ServerDisconnected += ServerDisconnected;
                client.Events.MessageReceived += MessageReceived;
                client.Callbacks.SyncRequestReceived = SyncRequestReceived;
                client.Connect();
                if (client.Connected)
                {

                    var rqType = new APITransport()
                    {
                        CommandCode = "CLIENT_TYPE",
                        TransportContent = "COM"
                    };
                    Thread.Sleep(50);
                    client.SendAsync(JsonConvert.SerializeObject(rqType));
                    if (args.Length > 0)
                    {
                        var rqTrans = new APITransport()
                        {
                            PartnerCode = args[0],
                            CommandCode = "COM_REQ_BILL",
                            TransportId = string.Empty,
                            TransportContent = JsonConvert.SerializeObject(new CashOut()
                            {
                                TransId = args[1],
                                MoId = args[2],
                                MoName = args[3],
                                Amount = Convert.ToInt32(args[4]),
                                Note = args[5],
                                CallbackUrl = args[6]
                            })
                        };
                        NLogLogger.Info(new string[] { "TCPCommand", "Main", "request Server", JsonConvert.SerializeObject(rqTrans) });
                        client.SendAsync(JsonConvert.SerializeObject(rqTrans));
                    }

                }
                Console.Write("SUCCESS");
            }
            catch (Exception exp)
            {
                Console.Write("FAIL");
                NLogLogger.Info(new string[] { "Main", "ERROR", exp.Message });
            }
        }

**Reply by jchristn** [reply]
You're using ```client.SendAsync``` and not either ```await```ing it or using ```.Wait()```.  The end of the app is reached before this task even starts.

**Reply by jchristn** [reply]
You can also use ```client.Send``` which is not an ```async``` method.

**Reply by binbon2d** [reply]
I understand synchronous and asynchronous
I used both of your suggestions
1.  client.SendAsync(JsonConvert.SerializeObject(rqTrans)).Wait();
2.  client.Send(JsonConvert.SerializeObject(rqTrans));

But no solve the problem

I try Send continuously, the server received but not the full number of count I send

**Reply by jchristn** [reply]
Got it.  I think I see what is happening after testing with the ```Test.FastDisconnect``` project.  There is an IO exception while it's trying to read the message (server-side).  I will investigate.  Thanks for bringing this to my attention!

**Reply by jchristn** [reply]
Ok, so this is actually a really hard problem to solve, but thankfully there are two viable solutions.  I tried hacking apart the logic for reading from a stream to actually return the data in the event of an exception (in case the message was fully read, you still want to process it).  That led to a rat's nest of problems.  

However, something I found which may be useful in your case is wrapping your instance of ```WatsonTcpClient``` in a ```using``` block.  That way, when it's done, it naturally calls ```.Dispose()```.  I tested this with the ```Test.FastDisconnect``` project, and was able to receive the message correctly.

Or, add a ```Task.Delay```, ```Thread.Sleep```, or something else.

**Reply by jchristn** [reply]
Example:
```
using System;
using WatsonTcp;

namespace Test.FastDisconnect
{
    class Program
    {
        static void Main(string[] args)
        {
            try
            {
                using (WatsonTcpClient client = new WatsonTcpClient("127.0.0.1", 9000))
                {
                    client.Events.MessageReceived += MessageReceived;
                    client.Connect();
                    client.Send("Hello!");
                }
            }
            catch (Exception e)
            {
                Console.WriteLine(e.ToString());
            }
        }

        static void MessageReceived(object sender, MessageReceivedEventArgs args)
        {

        }
    }
}
```

**Reply by binbon2d** [reply]
Thank you for your quick reply
I implemented according to your suggestion using `using` Block but the first time Star Server & Client still lost data
1. Start Server
2. Start Client & Connect
3. Client Send Message

If don't turn off the Server App,  next second sending everything works fine 

Temporarily this this problem does not affect my project too much, I will always follow you if there is a new version of WatsonTcp it has been helping me a lot for my work

Thank you very much

**Reply by binbon2d** [reply]
By the way, can you tell me if there is an **Event** on the **Server** side for **IdleClientTimeout** ?

I setting :
`server.Settings.IdleClientTimeoutSeconds = 0;`
or
`server.Settings.IdleClientTimeoutSeconds = Int32.MaxValue;`

The client does not communicate with the server for a period of time.
I feel that the Client has been TimeOut and the Server is not aware of this

I want my Client to always connect to the Server, what should I do?

**Reply by jchristn** [reply]
Hi @binbon2d sorry for the delay.  There is no event indicating a timeout, but the ```Events.ClientDisconnected``` event uses the ```DisconnectionEventArgs```, and ```Reason``` should be populated with ```Timeout```.  Please let me know if that's not the case.

I just tested using the ```Test.Server``` and ```Test.Client``` project, and the reason is indeed populated.

Client
```
C:\Code\Watson\WatsonTcp\src\Test.Client\bin\Debug\net7.0>test.client
Server IP: [localhost]
Server port: [9000]
Use SSL: [y/N]?
[Info     ] [WatsonTcpClient] connecting to localhost:9000
Server connected
[Info     ] [WatsonTcpClient] connected to localhost:9000
Command [? for help]: [Info     ] [WatsonTcpClient] disconnect due to timeout
[Debug    ] [WatsonTcpClient] data receiver terminated for localhost:9000
Server disconnected: Timeout
```

Server:
```
_Server.Settings.IdleClientTimeoutSeconds = 5;
```
